src/components/clusterexplorer/explorer.ts

```diff
@@ -156,7 +202,86 @@ export class KubernetesExplorer implements vscode.TreeDataProvider<ClusterExplor
         // and always queue a refresh.
         //
         // These are pretty niche cases, so I'm not too worried if they aren't perfect.
-        sleep(50).then(() => refreshExplorer());
+        if (!node) {
+            sleep(50).then(() => refreshExplorer());
+        } else {
+            const currentNodeId = this.getWatchId(node as ClusterExplorerNode);
+            if (!currentNodeId) {
+                return;
+            }
+            // In the case where many requests of updating are received in a short amount of time
+            // the tree is not refreshed for every change but once after a while.
+            // Every call resets a timer which trigger the tree refresh
+            clearTimeout(this.refreshTimer);
+            // check if element already is in refreshqueue before pushing it
+            const hasNode = this.refreshQueue.find((nodeQueue) => {
+                const nodeQueueid = this.getWatchId(nodeQueue);
+                return nodeQueueid && currentNodeId === nodeQueueid;
+            });
+            if (!hasNode) {
+                this.refreshQueue.push(node as ClusterExplorerNode);
+            }
+            this.refreshTimer = setTimeout(() => {
+                this.refreshQueue.splice(0).forEach((_node) => this.refresh(_node));
+            }, 500);
+        }
+    }
+
+    private onElementCollapsed(e: vscode.TreeViewExpansionEvent<ClusterExplorerNode>) {
+        this.collapse(e.element);
+	}
+
+	private onElementExpanded(e: vscode.TreeViewExpansionEvent<ClusterExplorerNode>) {
+        this.expand(e.element);
+    }
+
+    private expand(node: ClusterExplorerNode) {
+        const resourcesToWatch = getResourcesToBeWatched();
+        const watchId = this.getWatchId(node);
+        const treeItem = node.getTreeItem();
+        providerResult.transform(treeItem, (ti) => {
+            if (ti.collapsibleState === vscode.TreeItemCollapsibleState.Collapsed) {
+                ti.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+            }
+            if (watchId &&
+                (WatchManager.getInstance().existsWatch(watchId) ||
+                (ti.label &&
```

Comment:
You're right and i also had some doubts about it but in this case the label is only used to check if the user added that node to the resources he wants to watch. Let's suppose we change the code to deal with ids. The label is `nginx`, how could the user know which is the actual id for that node? By using labels we could end up creating some issues but it could also be seen as a feature.... if i want to watch both deployment and service called nginx, i type in `nginx` once and by expanding both nodes i can start watching them. WDYT?

Reply:
Ah, I see what you mean - `resourcesToWatch` is the config, so the text has to be something that a human stands a chance of typing.  IDs are definitely not the solution but `kubectl` references (`kind/name`) should be fairly stable and familiar.  But I see what you mean about the user possibly wanting to type `nginx` and not have to specify the kind; but then again I feel like this should go on resource name not on the node label itself.

How about something like this:

* bare name that is a known kind (`deployment`) - watch that resource type
* other bare name (`nginx`, `deployments`) - watch any node whose label, **or** resource node resource name, matches (case-insensitive because the UI title-cases folder names)
* name with slash (`deployment/nginx`) - watch only the node with that resource kind _and_ name

It is quite a bit fiddlier than what you have at the moment, though, so we could certainly defer it...

Reply:
Yes i think your example covers every scenario and it's quite clear and simple to be used by everyone +1